sw/device/silicon_creator/lib/drivers/keymgr.c

```diff
@@ -0,0 +1,95 @@
+
+#include "sw/device/silicon_creator/lib/drivers/keymgr.h"
+
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "keymgr_regs.h"  // Generated.
+
+static const uint32_t base = TOP_EARLGREY_KEYMGR_BASE_ADDR;
+
+/**
+ * Checks the key manager `expected_state`.
+ *
+ * This function reads and clears the status and error code registers.
+ *
+ * @return `kErrorOk` if the key manager is at the `expected_state` and the
+ * status is idle or success.
+ */
+static rom_error_t check_expected_state(uint32_t expected_state) {
+  // Read and clear the status register by writing back the read value.
+  uint32_t op_status = abs_mmio_read32(base + KEYMGR_OP_STATUS_REG_OFFSET);
+  abs_mmio_write32(base + KEYMGR_OP_STATUS_REG_OFFSET, op_status);
+
+  // Read and clear the error register by writing back the read value.
+  uint32_t error_code = abs_mmio_read32(base + KEYMGR_ERR_CODE_REG_OFFSET);
+  abs_mmio_write32(base + KEYMGR_ERR_CODE_REG_OFFSET, error_code);
+
+  uint32_t got_state = abs_mmio_read32(base + KEYMGR_WORKING_STATE_REG_OFFSET);
+  uint32_t op_status_field =
+      bitfield_field32_read(op_status, KEYMGR_OP_STATUS_STATUS_FIELD);
+  if ((op_status_field == KEYMGR_OP_STATUS_STATUS_VALUE_IDLE ||
+       op_status_field == KEYMGR_OP_STATUS_STATUS_VALUE_DONE_SUCCESS) &&
+      error_code == 0u && got_state == expected_state) {
+    return kErrorOk;
+  }
+  return kErrorKeymgrInternal;
+}
+
+/**
+ * Advances the sate of the key manager.
+ *
+ * The `check_expected_state()` function must be called before this function
+ * no ensure the key manager is ready to receive op commands.
+ */
+static void advance_state(void) {
+  uint32_t reg = bitfield_bit32_write(0, KEYMGR_CONTROL_START_BIT, true);
+  reg = bitfield_field32_write(reg, KEYMGR_CONTROL_DEST_SEL_FIELD,
+                               KEYMGR_CONTROL_DEST_SEL_VALUE_NONE);
+  reg = bitfield_field32_write(reg, KEYMGR_CONTROL_OPERATION_FIELD,
+                               KEYMGR_CONTROL_OPERATION_VALUE_ADVANCE);
+  abs_mmio_write32(base + KEYMGR_CONTROL_REG_OFFSET, reg);
+}
+
+rom_error_t keymgr_init(uint16_t entropy_reseed_interval) {
+  RETURN_IF_ERROR(check_expected_state(KEYMGR_WORKING_STATE_STATE_VALUE_RESET));
+
+  uint32_t reg = bitfield_field32_write(0, KEYMGR_RESEED_INTERVAL_VAL_FIELD,
+                                        entropy_reseed_interval);
+  abs_mmio_write32(base + KEYMGR_RESEED_INTERVAL_REG_OFFSET, reg);
+
+  // Advance to INIT state.
+  advance_state();
+  return kErrorOk;
+}
+
+rom_error_t keymgr_advance_creator_state(const uint32_t binding_value[8],
+                                         uint32_t max_key_ver) {
+  RETURN_IF_ERROR(check_expected_state(KEYMGR_WORKING_STATE_STATE_VALUE_INIT));
+
+  // Write and lock (rw0c) the software binding value. This register is unlocked
+  // by hardware upon a successful state transition.
+  // FIXME: Consider using sec_mmio module for the following register writes.
```

Comment:
I think we should definitely do this, probably for all applicable writes of this driver. If not, I guess at a minimum this function should check that registers are not already locked.